### PR TITLE
cryptsetup: convert pkcs11/fido2 to iovec for key handling

### PR DESCRIPTION
**Pull request by bluca**
key-data might be NULL. Fixes crash:

```
0  0x0000559c62120530 in attach_luks_or_plain_or_bitlk (cd=0x559c6b192830, name=0x7ffd57981dc4 "root", token_type=TOKEN_FIDO2, key_file=0x0, key_data=0x0, passwords=0x0, flags=524296, until=0)
    at ../src/cryptsetup/cryptsetup.c:2234
        pass_volume_key = false
        r = 1469577760
        __func__ = '\000' <repeats 29 times>
1  0x0000559c6212279c in run (argc=6, argv=0x7ffd5797fe98) at ../src/cryptsetup/cryptsetup.c:2597
        discovered_key_data = {iov_base = 0x0, iov_len = 0}
        key_data = 0x0
        token_type = TOKEN_FIDO2
        destroy_key_file = 0x0
        flags = 524296
        until = 0
        passphrase_type = PASSPHRASE_NONE
        volume = 0x7ffd57981dc4 "root"
        source = 0x7ffd57981dc9 "/dev/disk/by-uuid/8372fb39-9ba4-461a-a618-07dcaae66280"
        status = CRYPT_INACTIVE
        tries = 0
        key_file = 0x0
        config = 0x7ffd57981e05 "luks,discard,fido2-device=auto,x-initrd.attach"
        use_cached_passphrase = true
        try_discover_key = true
        discovered_key_fn = 0x7ffd5797fa70 "root.key"
        passwords = 0x0
        cd = 0x559c6b192830
        verb = 0x7ffd57981dbd "attach"
        r = 0
        __func__ = "\000\000\000"
2  0x0000559c621231e6 in main (argc=6, argv=0x7ffd5797fe98) at ../src/cryptsetup/cryptsetup.c:2674
        r = 32553
        __func__ = "\000\000\000\000"
```

Follow-up for 53b6c99018f918a5d2c9000ac5fe3a2440115ea7

### CODE REVIEW

File: src/shared/cryptsetup-fido2.c [hunk 1]
```diff
@@ -252,7 +251,7 @@ int acquire_fido2_key_auto(
                                 /* key_file= */ NULL, /* salt is read from LUKS header instead of key_file */
                                 /* key_file_size= */ 0,
                                 /* key_file_offset= */ 0,
-                                salt, salt_size,
+                                &(struct iovec) { salt, salt_size },
```
Comment (YHNdnzj) [inline src/shared/cryptsetup-fido2.c]: &IOVEC_MAKE()

File: src/cryptsetup/cryptsetup-pkcs11.c [hunk 2]
```diff
@@ -31,8 +32,7 @@ int decrypt_pkcs11_key(
                 const char *key_file,         /* We either expect key_file and associated parameters to be set (for file keys) … */
                 size_t key_file_size,
                 uint64_t key_file_offset,
-                const void *key_data,         /* … or key_data and key_data_size (for literal keys) */
-                size_t key_data_size,
+                const struct iovec *key_data, /* … or key_data and key_data_size (for literal keys) */
```
Comment (YHNdnzj) [inline src/cryptsetup/cryptsetup-pkcs11.c]: "or literal keys via key_data"?

**Review by YHNdnzj — APPROVED** [review]

File: src/shared/cryptsetup-fido2.c [hunk 3]
```diff
@@ -45,10 +44,10 @@ int acquire_fido2_key(
                                         "Local verification is required to unlock this volume, but the 'headless' parameter was set.");
 
         assert(cid);
-        assert(key_file || key_data);
+        assert(key_file || iovec_is_set(key_data));
 
-        if (key_data)
-                salt = IOVEC_MAKE(key_data, key_data_size);
+        if (iovec_is_set(key_data))
+                salt = IOVEC_MAKE(key_data->iov_base, key_data->iov_len);
```
Comment (yuwata) [inline src/shared/cryptsetup-fido2.c]:
```c
salt = *key_data;
```
is enough?

**Comment by yuwata** [comment]
BTW, The code path seems never tested. Is it possible to add a test case?

**Comment by bluca** [comment]
> BTW, The code path seems never tested. Is it possible to add a test case?

It's a lot of work to mock pkcs11/fido2, so theorethically possible, but very difficult in practice

**Pull request merged**